From d5a7badd6a0ea4ecbe76f0205aa53b42f7cd90dc Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Mon, 5 May 2008 14:37:07 -0400
Subject: [PATCH] Fix hal shutdown crash.

Removing the device invalidates its ->next pointer.  Copy it aside before
destroying the device.
(cherry picked from commit f52f6c5c7efc281f9ac204fbaa4f71383df7463d)
---
 config/hal.c |    8 +++++---
 1 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/config/hal.c b/config/hal.c
index 16f16ec..bdf7d6c 100644
--- a/config/hal.c
+++ b/config/hal.c
@@ -63,7 +63,7 @@ remove_device(DeviceIntPtr dev)
 static void
 device_removed(LibHalContext *ctx, const char *udi)
 {
-    DeviceIntPtr dev;
+    DeviceIntPtr dev, next;
     char *value;
 
     value = xalloc(strlen(udi) + 5); /* "hal:" + NULL */
@@ -71,11 +71,13 @@ device_removed(LibHalContext *ctx, const char *udi)
         return;
     sprintf(value, "hal:%s", udi);
 
-    for (dev = inputInfo.devices; dev; dev = dev->next) {
+    for (dev = inputInfo.devices; dev; dev = next) {
+	next = dev->next;
         if (dev->config_info && strcmp(dev->config_info, value) == 0)
             remove_device(dev);
     }
-    for (dev = inputInfo.off_devices; dev; dev = dev->next) {
+    for (dev = inputInfo.off_devices; dev; dev = next) {
+	next = dev->next;
         if (dev->config_info && strcmp(dev->config_info, value) == 0)
             remove_device(dev);
     }
-- 
1.5.5.1

